diff -Naurp gcc/c-decl.c gcc/c-decl.c
--- gcc/c-decl.c	2012-02-23 23:33:33.000000000 +0530
+++ gcc/c-decl.c	2013-03-18 10:04:59.000000000 +0530
@@ -2463,8 +2463,8 @@ undeclared_variable (tree id, location_t
 
       if (!already)
 	{
-	  error ("%H(Each undeclared identifier is reported only once", &loc);
-	  error ("%Hfor each function it appears in.)", &loc);
+          inform (loc, "each undeclared identifier is reported only"
+                  " once for each function it appears in");
 	  already = true;
 	}
 
