diff -Naurp gcc/config/avr32/avr32.c gcc/config/avr32/avr32.c
--- gcc/config/avr32/avr32.c	2011-10-07 16:58:16.000000000 +0530
+++ gcc/config/avr32/avr32.c	2011-10-07 15:37:07.000000000 +0530
@@ -167,6 +167,8 @@ static const struct part_type_s avr32_pa
   {"uc3c2256c",       PART_TYPE_AVR32_UC3C2256C,       ARCH_TYPE_AVR32_UCR3FP,    "__AVR32_UC3C2256C__"},
   {"uc3c2512c",       PART_TYPE_AVR32_UC3C2512C,       ARCH_TYPE_AVR32_UCR3FP,    "__AVR32_UC3C2512C__"},
   {"mxt768e",         PART_TYPE_AVR32_MXT768E,         ARCH_TYPE_AVR32_UCR3,      "__AVR32_MXT768E__"},
+  {"mxt1664s",        PART_TYPE_AVR32_MXT1664S,        ARCH_TYPE_AVR32_UCR3,      "__AVR32_MXT1664S__"},
+  {"mxt1664eng",      PART_TYPE_AVR32_MXT1664ENG,      ARCH_TYPE_AVR32_UCR3,      "__AVR32_MXT1664ENG__"},
   {NULL, 0, 0, NULL}
 };
 
diff -Naurp gcc/config/avr32/avr32.h gcc/config/avr32/avr32.h
--- gcc/config/avr32/avr32.h	2011-10-07 16:58:16.000000000 +0530
+++ gcc/config/avr32/avr32.h	2011-10-07 15:38:21.000000000 +0530
@@ -219,7 +219,9 @@ enum part_type
   PART_TYPE_AVR32_UC3C2128C,
   PART_TYPE_AVR32_UC3C2256C,
   PART_TYPE_AVR32_UC3C2512C,
-  PART_TYPE_AVR32_MXT768E
+  PART_TYPE_AVR32_MXT768E,
+  PART_TYPE_AVR32_MXT1664S,
+  PART_TYPE_AVR32_MXT1664ENG
 };
 
 /* Microarchitectures. */
diff -Naurp gcc/config/avr32/t-avr32 gcc/config/avr32/t-avr32
--- gcc/config/avr32/t-avr32	2011-10-07 16:58:16.000000000 +0530
+++ gcc/config/avr32/t-avr32	2011-10-07 15:39:44.000000000 +0530
@@ -96,6 +96,8 @@ MULTILIB_MATCHES     += march?ucr3fp=mpa
 MULTILIB_MATCHES     += march?ucr3fp=mpart?uc3c2256c
 MULTILIB_MATCHES     += march?ucr3fp=mpart?uc3c2512c
 MULTILIB_MATCHES     += march?ucr3=mpart?mxt768e
+MULTILIB_MATCHES     += march?ucr3=mpart?mxt1664s
+MULTILIB_MATCHES     += march?ucr3=mpart?mxt1664eng
 
 
 EXTRA_MULTILIB_PARTS = crtbegin.o crtbeginS.o crtend.o crtendS.o crti.o crtn.o
diff -Naurp gcc/config/avr32/t-avr32-linux gcc/config/avr32/t-avr32-linux
--- gcc/config/avr32/t-avr32-linux	2011-10-07 16:58:16.000000000 +0530
+++ gcc/config/avr32/t-avr32-linux	2011-10-07 15:40:40.000000000 +0530
@@ -96,6 +96,8 @@ MULTILIB_MATCHES     += march?ucr3fp=mpa
 MULTILIB_MATCHES     += march?ucr3fp=mpart?uc3c2256c
 MULTILIB_MATCHES     += march?ucr3fp=mpart?uc3c2512c
 MULTILIB_MATCHES     += march?ucr3=mpart?mxt768e
+MULTILIB_MATCHES     += march?ucr3=mpart?mxt1664s
+MULTILIB_MATCHES     += march?ucr3=mpart?mxt1664eng
 
 
 EXTRA_MULTILIB_PARTS = crtbegin.o crtbeginS.o crtend.o crtendS.o
diff -Naurp gcc/doc/invoke.texi gcc/doc/invoke.texi
--- gcc/doc/invoke.texi	2011-10-07 16:58:16.000000000 +0530
+++ gcc/doc/invoke.texi	2011-10-07 15:42:10.000000000 +0530
@@ -9450,7 +9450,9 @@ Generate code for the specified part. Pe
 @samp{uc3c2128c},
 @samp{uc3c2256c},
 @samp{uc3c2512c},
-@samp{mxt768e}.
+@samp{mxt768e},
+@samp{mxt1664s},
+@samp{mxt1664eng}.
 
 @item -mcpu=@var{cpu-type}
 @opindex mcpu
