diff -Naurp gcc/config/avr32/avr32.c gcc/config/avr32/avr32.c
--- gcc/config/avr32/avr32.c	2012-11-26 12:01:08.000000000 +0530
+++ gcc/config/avr32/avr32.c	2012-11-26 12:08:15.000000000 +0530
@@ -167,6 +167,7 @@ static const struct part_type_s avr32_pa
   {"uc3c2128c",       PART_TYPE_AVR32_UC3C2128C,       ARCH_TYPE_AVR32_UCR3FP,    "__AVR32_UC3C2128C__"},
   {"uc3c2256c",       PART_TYPE_AVR32_UC3C2256C,       ARCH_TYPE_AVR32_UCR3FP,    "__AVR32_UC3C2256C__"},
   {"uc3c2512c",       PART_TYPE_AVR32_UC3C2512C,       ARCH_TYPE_AVR32_UCR3FP,    "__AVR32_UC3C2512C__"},
+  {"mxt2952t",        PART_TYPE_AVR32_MXT2952T,        ARCH_TYPE_AVR32_UCR3,      "__AVR32_MXT2952T__"},
   {"mxt540e",         PART_TYPE_AVR32_MXT540E,         ARCH_TYPE_AVR32_UCR3,      "__AVR32_MXT540E__"},
   {"mxt768e",         PART_TYPE_AVR32_MXT768E,         ARCH_TYPE_AVR32_UCR3,      "__AVR32_MXT768E__"},
   {"mxt1664s",        PART_TYPE_AVR32_MXT1664S,        ARCH_TYPE_AVR32_UCR3,      "__AVR32_MXT1664S__"},
diff -Naurp gcc/config/avr32/avr32.h gcc/config/avr32/avr32.h
--- gcc/config/avr32/avr32.h	2012-11-26 12:01:08.000000000 +0530
+++ gcc/config/avr32/avr32.h	2012-11-26 12:09:38.000000000 +0530
@@ -220,6 +220,7 @@ enum part_type
   PART_TYPE_AVR32_UC3C2128C,
   PART_TYPE_AVR32_UC3C2256C,
   PART_TYPE_AVR32_UC3C2512C,
+  PART_TYPE_AVR32_MXT2952T,
   PART_TYPE_AVR32_MXT540E,
   PART_TYPE_AVR32_MXT768E,
   PART_TYPE_AVR32_MXT1664S,
diff -Naurp gcc/config/avr32/t-avr32 gcc/config/avr32/t-avr32
--- gcc/config/avr32/t-avr32	2012-11-26 12:01:07.000000000 +0530
+++ gcc/config/avr32/t-avr32	2012-11-26 12:12:20.000000000 +0530
@@ -96,6 +96,7 @@ MULTILIB_MATCHES     += march?ucr3fp=mpa
 MULTILIB_MATCHES     += march?ucr3fp=mpart?uc3c2128c
 MULTILIB_MATCHES     += march?ucr3fp=mpart?uc3c2256c
 MULTILIB_MATCHES     += march?ucr3fp=mpart?uc3c2512c
+MULTILIB_MATCHES     += march?ucr3=mpart?mxt2952t
 MULTILIB_MATCHES     += march?ucr3=mpart?mxt540e
 MULTILIB_MATCHES     += march?ucr3=mpart?mxt768e
 MULTILIB_MATCHES     += march?ucr3=mpart?mxt1664s
diff -Naurp gcc/config/avr32/t-avr32-linux gcc/config/avr32/t-avr32-linux
--- gcc/config/avr32/t-avr32-linux	2012-11-26 12:01:08.000000000 +0530
+++ gcc/config/avr32/t-avr32-linux	2012-11-26 12:14:15.000000000 +0530
@@ -96,6 +96,7 @@ MULTILIB_MATCHES     += march?ucr3fp=mpa
 MULTILIB_MATCHES     += march?ucr3fp=mpart?uc3c2128c
 MULTILIB_MATCHES     += march?ucr3fp=mpart?uc3c2256c
 MULTILIB_MATCHES     += march?ucr3fp=mpart?uc3c2512c
+MULTILIB_MATCHES     += march?ucr3=mpart?mxt2952t
 MULTILIB_MATCHES     += march?ucr3=mpart?mxt540e
 MULTILIB_MATCHES     += march?ucr3=mpart?mxt768e
 MULTILIB_MATCHES     += march?ucr3=mpart?mxt1664s
diff -Naurp gcc/doc/invoke.texi gcc/doc/invoke.texi
--- gcc/doc/invoke.texi	2012-11-26 12:01:11.000000000 +0530
+++ gcc/doc/invoke.texi	2012-11-26 12:15:15.000000000 +0530
@@ -9451,6 +9451,7 @@ Generate code for the specified part. Pe
 @samp{uc3c2128c},
 @samp{uc3c2256c},
 @samp{uc3c2512c},
+@samp{mxt2952t},
 @samp{mxt540e},
 @samp{mxt768e},
 @samp{mxt1664s},
